File: 5_december/paringshoes.js
/**
 * @param {{ type: 'I' | 'R', size: number }[]} shoes
 * @returns {number[]} Available shoes
 */
function organizeShoes(shoes) {
    const sizeMap = new Map();

    // Contar botas izquierdas ('I') y derechas ('R') por tamaño
    for (const { type, size } of shoes) {
        if (!sizeMap.has(size)) {
            sizeMap.set(size, { I: 0, R: 0 });
        }
        sizeMap.get(size)[type]++;
    }

    const result = [];

    // Calcular pares y agregarlos al resultado
    for (const [size, counts] of sizeMap) {
        const pairs = Math.min(counts.I, counts.R); // Contar los pares disponibles
        result.push(...Array(pairs).fill(size));
    }

    return result;
}

const shoes = [
    { type: 'I', size: 38 },
    { type: 'R', size: 38 },
    { type: 'I', size: 38 },
    { type: 'I', size: 38 },
    { type: 'R', size: 38 }
]

console.log(organizeShoes(shoes))

File: 5_december/paringshoes.py
def organizeShoes(shoes):
  ordered_shoes = {}
  for shoe in shoes:
    if shoe['size'] in ordered_shoes:
      if shoe['type'] in ordered_shoes[shoe['size']]:
        ordered_shoes[shoe['size']][shoe['type']] += 1
      else:
        ordered_shoes[shoe['size']][shoe['type']] = 1
    else:
      ordered_shoes[shoe['size']] = {}
      ordered_shoes[shoe['size']][shoe['type']] = 1
      if shoe['type'] == 'R':
        ordered_shoes[shoe['size']]['I'] = 0
      else:
        ordered_shoes[shoe['size']]['R'] = 0

  response = []
  for size in ordered_shoes:
    pairs = (ordered_shoes[size]['I'] + ordered_shoes[size]['R'])/2
    if int(pairs) > 0:
      response += [size for _ in range(int(pairs))]

  return response

shoes = [
  { 'type': 'I', 'size': 38 },
  { 'type': 'R', 'size': 38 },
  { 'type': 'R', 'size': 42 },
  { 'type': 'I', 'size': 41 },
  { 'type': 'I', 'size': 42 }
]
ordered_shoes = organizeShoes(shoes)
print(ordered_shoes)